File: trunk/open-dis/Csharp/DIS/RecordQueryReliablePdu.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Diagnostics;
using System.Xml.Serialization;
using DISnet.DataStreamUtilities;

namespace DIS1998net
{

/**
 * Section 5.3.12.13: A request for one or more records of data from an entity. COMPLETE
 *
 * Copyright (c) 2008, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 * Modified for use with C#:
 * Peter Smith (Naval Air Warfare Center - Training Systems Division)
 */
[Serializable]
[XmlRoot]
[XmlInclude(typeof(FourByteChunk))]
public class RecordQueryReliablePdu : SimulationManagementWithReliabilityFamilyPdu
{
   /** request ID */
   protected uint  _requestID;

   /** level of reliability service used for this transaction */
   protected byte  _requiredReliabilityService;

   /** padding. The spec is unclear and contradictory here. */
   protected ushort  _pad1;

   /** padding */
   protected byte  _pad2;

   /** event type */
   protected ushort  _eventType;

   /** time */
   protected uint  _time;

   /** numberOfRecords */
   protected uint  _numberOfRecords;

   /** record IDs */
   protected List<object> _recordIDs = new List<object>(); 

/** Constructor */
 public RecordQueryReliablePdu()
 {
    PduType = (byte)63;
 }

public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = base.getMarshalledSize();
   marshalSize = marshalSize + 4;  // _requestID
   marshalSize = marshalSize + 1;  // _requiredReliabilityService
   marshalSize = marshalSize + 2;  // _pad1
   marshalSize = marshalSize + 1;  // _pad2
   marshalSize = marshalSize + 2;  // _eventType
   marshalSize = marshalSize + 4;  // _time
   marshalSize = marshalSize + 4;  // _numberOfRecords
   for(int idx=0; idx < _recordIDs.Count; idx++)
   {
        FourByteChunk listElement = (FourByteChunk)_recordIDs[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setRequestID(uint pRequestID)
{ _requestID = pRequestID;
}

[XmlElement(Type= typeof(uint), ElementName="requestID")]
public uint RequestID
{
     get
{
          return _requestID;
}
     set
{
          _requestID = value;
}
}

public void setRequiredReliabilityService(byte pRequiredReliabilityService)
{ _requiredReliabilityService = pRequiredReliabilityService;
}

[XmlElement(Type= typeof(byte), ElementName="requiredReliabilityService")]
public byte RequiredReliabilityService
{
     get
{
          return _requiredReliabilityService;
}
     set
{
          _requiredReliabilityService = value;
}
}

public void setPad1(ushort pPad1)
{ _pad1 = pPad1;
}

[XmlElement(Type= typeof(ushort), ElementName="pad1")]
public ushort Pad1
{
     get
{
          return _pad1;
}
     set
{
          _pad1 = value;
}
}

public void setPad2(byte pPad2)
{ _pad2 = pPad2;
}

[XmlElement(Type= typeof(byte), ElementName="pad2")]
public byte Pad2
{
     get
{
          return _pad2;
}
     set
{
          _pad2 = value;
}
}

public void setEventType(ushort pEventType)
{ _eventType = pEventType;
}

[XmlElement(Type= typeof(ushort), ElementName="eventType")]
public ushort EventType
{
     get
{
          return _eventType;
}
     set
{
          _eventType = value;
}
}

public void setTime(uint pTime)
{ _time = pTime;
}

[XmlElement(Type= typeof(uint), ElementName="time")]
public uint Time
{
     get
{
          return _time;
}
     set
{
          _time = value;
}
}

public void setRecordIDs(List<object> pRecordIDs)
{ _recordIDs = pRecordIDs;
}

public List<object> getRecordIDs()
{ return _recordIDs; }

[XmlElement(ElementName = "recordIDsList",Type = typeof(List<object>))]
public List<object> RecordIDs
{
     get
{
          return _recordIDs;
}
     set
{
          _recordIDs = value;
}
}


public void marshal(DataOutputStream dos)
{
    base.marshal(dos);
    try 
    {
       dos.writeUint( (uint)_requestID);
       dos.writeByte( (byte)_requiredReliabilityService);
       dos.writeUshort( (ushort)_pad1);
       dos.writeByte( (byte)_pad2);
       dos.writeUshort( (ushort)_eventType);
       dos.writeUint( (uint)_time);
       dos.writeUint( (uint)_recordIDs.Count);

       for(int idx = 0; idx < _recordIDs.Count; idx++)
       {
            FourByteChunk aFourByteChunk = (FourByteChunk)_recordIDs[idx];
            aFourByteChunk.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      Trace.WriteLine(e);
      Trace.Flush();
    }
} // end of marshal method

public void unmarshal(DataInputStream dis)
{
    base.unmarshal(dis);

    try 
    {
       _requestID = dis.readUint();
       _requiredReliabilityService = dis.readByte();
       _pad1 = dis.readUshort();
       _pad2 = dis.readByte();
       _eventType = dis.readUshort();
       _time = dis.readUint();
       _numberOfRecords = dis.readUint();
        for(int idx = 0; idx < _numberOfRecords; idx++)
        {
           FourByteChunk anX = new FourByteChunk();
            anX.unmarshal(dis);
            _recordIDs.Add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      Trace.WriteLine(e); 
      Trace.Flush();
    }
 } // end of unmarshal method 


public void reflection(StringBuilder sb)
{
    sb.Append("----- RecordQueryReliablePdu-----"  + System.Environment.NewLine);
    base.reflection(sb);
    try 
    {
           sb.Append("uint\t _requestID\t " + _requestID.ToString() + System.Environment.NewLine);
           sb.Append("byte\t _requiredReliabilityService\t " + _requiredReliabilityService.ToString() + System.Environment.NewLine);
           sb.Append("ushort\t _pad1\t " + _pad1.ToString() + System.Environment.NewLine);
           sb.Append("byte\t _pad2\t " + _pad2.ToString() + System.Environment.NewLine);
           sb.Append("ushort\t _eventType\t " + _eventType.ToString() + System.Environment.NewLine);
           sb.Append("uint\t _time\t " + _time.ToString() + System.Environment.NewLine);
           sb.Append("uint\t _recordIDs\t " + _recordIDs.Count.ToString() + System.Environment.NewLine);

       for(int idx = 0; idx < _recordIDs.Count; idx++)
       {
           sb.Append("FourByteChunk\t " + _recordIDs[idx] + System.Environment.NewLine);
            FourByteChunk aFourByteChunk = (FourByteChunk)_recordIDs[idx];
            aFourByteChunk.reflection(sb);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      Trace.WriteLine(e);
      Trace.Flush();
}
    } // end of marshal method

 /**
  * The equals method doesn't always work--mostly it works only on on classes that consist only of primitives. Be careful.
  */
 public bool equals(RecordQueryReliablePdu rhs)
 {
     bool ivarsEqual = true;

    if(rhs.GetType() != this.GetType())
        return false;

     if( ! (_requestID == rhs._requestID)) ivarsEqual = false;
     if( ! (_requiredReliabilityService == rhs._requiredReliabilityService)) ivarsEqual = false;
     if( ! (_pad1 == rhs._pad1)) ivarsEqual = false;
     if( ! (_pad2 == rhs._pad2)) ivarsEqual = false;
     if( ! (_eventType == rhs._eventType)) ivarsEqual = false;
     if( ! (_time == rhs._time)) ivarsEqual = false;
     if( ! (_numberOfRecords == rhs._numberOfRecords)) ivarsEqual = false;

     for(int idx = 0; idx < _recordIDs.Count; idx++)
     {
        FourByteChunk x = (FourByteChunk)_recordIDs[idx];
        if( ! ( _recordIDs[idx].Equals(rhs._recordIDs[idx]))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class
} // end of namespace

File: trunk/open-dis/java/edu/nps/moves/dis/RecordQueryReliablePdu.java
package edu.nps.moves.dis;

import java.util.*;
import java.io.*;
import javax.xml.bind.annotation.*;

/**
 * Section 5.3.12.13: A request for one or more records of data from an entity. COMPLETE
 *
 * Copyright (c) 2008, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 * @version $Id:$
 */
public class RecordQueryReliablePdu extends SimulationManagementWithReliabilityFamilyPdu implements Serializable {

    /** request ID */
    protected long requestID;
    /** level of reliability service used for this transaction */
    protected short requiredReliabilityService;
    /** padding. The spec is unclear and contradictory here. */
    protected int pad1;
    /** padding */
    protected short pad2;
    /** event type */
    protected int eventType;
    /** time */
    protected long time;
    /** numberOfRecords */
    protected long numberOfRecords;
    /** record IDs */
    protected List<FourByteChunk> recordIDs = new ArrayList<FourByteChunk>();

    /** Constructor */
    public RecordQueryReliablePdu() {
        setPduType((short) 63);
    }

    @Override
    public int getMarshalledSize() {
        int marshalSize = 0;

        marshalSize = super.getMarshalledSize();
        marshalSize = marshalSize + 4;  // requestID
        marshalSize = marshalSize + 1;  // requiredReliabilityService
        marshalSize = marshalSize + 2;  // pad1
        marshalSize = marshalSize + 1;  // pad2
        marshalSize = marshalSize + 2;  // eventType
        marshalSize = marshalSize + 4;  // time
        marshalSize = marshalSize + 4;  // numberOfRecords
        for (int idx = 0; idx < recordIDs.size(); idx++) {
            FourByteChunk listElement = recordIDs.get(idx);
            marshalSize = marshalSize + listElement.getMarshalledSize();
        }

        return marshalSize;
    }

    public void setRequestID(long pRequestID) {
        requestID = pRequestID;
    }

    @XmlAttribute
    public long getRequestID() {
        return requestID;
    }

    public void setRequiredReliabilityService(short pRequiredReliabilityService) {
        requiredReliabilityService = pRequiredReliabilityService;
    }

    @XmlAttribute
    public short getRequiredReliabilityService() {
        return requiredReliabilityService;
    }

    public void setPad1(int pPad1) {
        pad1 = pPad1;
    }

    @XmlAttribute
    public int getPad1() {
        return pad1;
    }

    public void setPad2(short pPad2) {
        pad2 = pPad2;
    }

    @XmlAttribute
    public short getPad2() {
        return pad2;
    }

    public void setEventType(int pEventType) {
        eventType = pEventType;
    }

    @XmlAttribute
    public int getEventType() {
        return eventType;
    }

    public void setTime(long pTime) {
        time = pTime;
    }

    @XmlAttribute
    public long getTime() {
        return time;
    }

    @XmlAttribute
    public long getNumberOfRecords() {
        return (long) recordIDs.size();
    }

    /** Note that setting this value will not change the marshalled value. The list whose length this describes is used for that purpose.
     * The getnumberOfRecords method will also be based on the actual list length rather than this value.
     * The method is simply here for java bean completeness.
     * @param pNumberOfRecords
     */
    public void setNumberOfRecords(long pNumberOfRecords) {
        numberOfRecords = pNumberOfRecords;
    }

    public void setRecordIDs(List<FourByteChunk> pRecordIDs) {
        recordIDs = pRecordIDs;
    }

    @XmlElementWrapper(name = "recordIDsList")
    public List<FourByteChunk> getRecordIDs() {
        return recordIDs;
    }

    @Override
    public void marshal(DataOutputStream dos) {
        super.marshal(dos);
        try {
            dos.writeInt((int) requestID);
            dos.writeByte((byte) requiredReliabilityService);
            dos.writeShort((short) pad1);
            dos.writeByte((byte) pad2);
            dos.writeShort((short) eventType);
            dos.writeInt((int) time);
            dos.writeInt(recordIDs.size());

            for (int idx = 0; idx < recordIDs.size(); idx++) {
                FourByteChunk aFourByteChunk = recordIDs.get(idx);
                aFourByteChunk.marshal(dos);
            } // end of list marshalling

        } // end try
        catch (Exception e) {
            System.out.println(e);
        }
    } // end of marshal method

    @Override
    public void unmarshal(DataInputStream dis) {
        super.unmarshal(dis);

        try {
            requestID = dis.readInt();
            requiredReliabilityService = (short) dis.readUnsignedByte();
            pad1 = dis.readUnsignedShort();
            pad2 = (short) dis.readUnsignedByte();
            eventType = dis.readUnsignedShort();
            time = dis.readInt();
            numberOfRecords = dis.readInt();
            for (int idx = 0; idx < numberOfRecords; idx++) {
                FourByteChunk anX = new FourByteChunk();
                anX.unmarshal(dis);
                recordIDs.add(anX);
            }

        } // end try
        catch (Exception e) {
            System.out.println(e);
        }
    } // end of unmarshal method

    /**
     * Packs a Pdu into the ByteBuffer.
     * @throws java.nio.BufferOverflowException if buff is too small
     * @throws java.nio.ReadOnlyBufferException if buff is read only
     * @see java.nio.ByteBuffer
     * @param buff The ByteBuffer at the position to begin writing
     * @since ??
     */
    @Override
    public void marshal(java.nio.ByteBuffer buff) {
        super.marshal(buff);
        buff.putInt((int) requestID);
        buff.put((byte) requiredReliabilityService);
        buff.putShort((short) pad1);
        buff.put((byte) pad2);
        buff.putShort((short) eventType);
        buff.putInt((int) time);
        buff.putInt(recordIDs.size());

        for (int idx = 0; idx < recordIDs.size(); idx++) {
            FourByteChunk aFourByteChunk = recordIDs.get(idx);
            aFourByteChunk.marshal(buff);
        } // end of list marshalling

    } // end of marshal method

    /**
     * Unpacks a Pdu from the underlying data.
     * @throws java.nio.BufferUnderflowException if buff is too small
     * @see java.nio.ByteBuffer
     * @param buff The ByteBuffer at the position to begin reading
     * @since ??
     */
    @Override
    public void unmarshal(java.nio.ByteBuffer buff) {
        super.unmarshal(buff);

        requestID = buff.getInt();
        requiredReliabilityService = (short) (buff.get() & 0xFF);
        pad1 = (buff.getShort() & 0xFFFF);
        pad2 = (short) (buff.get() & 0xFF);
        eventType = (buff.getShort() & 0xFFFF);
        time = buff.getInt();
        numberOfRecords = buff.getInt();
        for (int idx = 0; idx < numberOfRecords; idx++) {
            FourByteChunk anX = new FourByteChunk();
            anX.unmarshal(buff);
            recordIDs.add(anX);
        }

    } // end of unmarshal method

    /**
     * The equals method doesn't always work--mostly on on classes that consist only of primitives. Be careful.
     * @param rhs
     * @return
     */
    public boolean equals(RecordQueryReliablePdu rhs) {
        boolean ivarsEqual = true;

        if (rhs.getClass() != this.getClass()) {
            return false;
        }

        if (!(requestID == rhs.requestID)) {
            ivarsEqual = false;
        }
        if (!(requiredReliabilityService == rhs.requiredReliabilityService)) {
            ivarsEqual = false;
        }
        if (!(pad1 == rhs.pad1)) {
            ivarsEqual = false;
        }
        if (!(pad2 == rhs.pad2)) {
            ivarsEqual = false;
        }
        if (!(eventType == rhs.eventType)) {
            ivarsEqual = false;
        }
        if (!(time == rhs.time)) {
            ivarsEqual = false;
        }
        if (!(numberOfRecords == rhs.numberOfRecords)) {
            ivarsEqual = false;
        }

        for (int idx = 0; idx < recordIDs.size(); idx++) {
            FourByteChunk x = recordIDs.get(idx);
            if (!(x.equals(rhs.recordIDs.get(idx)))) {
                ivarsEqual = false;
            }
        }

        return ivarsEqual;
    }
} // end of class
